mongorestore/filepath.go

```diff
@@ -204,41 +204,105 @@ func (f *stdinFile) Close() error {
-func (restore *MongoRestore) getInfoFromFilename(filename string) (string, FileType, error) {
+// getInfoFromFile returns the collection name and FileType from a bson or metadata file.
+// The collection name may be pulled from either the file name itself, or from the content
+// of a .metadata.json file if the file name is truncated.
+func (restore *MongoRestore) getInfoFromFile(filename string) (string, FileType, error) {
 	baseFileName := filepath.Base(filename)
+	metadataFullPath := ""
+
+	collName := ""
+	unescapedCollName := ""
 
-	baseName := ""
 	fileType := UnknownFileType
+	var err error
 
 	// .bin supported for legacy reasons
 	if strings.HasSuffix(baseFileName, ".bin") {
-		baseName = strings.TrimSuffix(baseFileName, ".bin")
+		collName = strings.TrimSuffix(baseFileName, ".bin")
 		fileType = BSONFileType
 	} else if restore.InputOptions.Gzip && restore.InputOptions.Archive == "" {
 		// Gzip indicates that files in a dump directory should have a .gz suffix
 		// but it does not indicate that the "files" provided by the archive should,
 		// compressed or otherwise.
 		if strings.HasSuffix(baseFileName, ".metadata.json.gz") {
-			baseName = strings.TrimSuffix(baseFileName, ".metadata.json.gz")
+			collName = strings.TrimSuffix(baseFileName, ".metadata.json.gz")
 			fileType = MetadataFileType
+			metadataFullPath = filename
 		} else if strings.HasSuffix(baseFileName, ".bson.gz") {
-			baseName = strings.TrimSuffix(baseFileName, ".bson.gz")
+			collName = strings.TrimSuffix(baseFileName, ".bson.gz")
 			fileType = BSONFileType
+			metadataFullPath = strings.TrimSuffix(filename, ".bson.gz") + ".metadata.json.gz"
 		}
 	} else if strings.HasSuffix(baseFileName, ".metadata.json") {
-		baseName = strings.TrimSuffix(baseFileName, ".metadata.json")
+		collName = strings.TrimSuffix(baseFileName, ".metadata.json")
 		fileType = MetadataFileType
+		metadataFullPath = filename
 	} else if strings.HasSuffix(baseFileName, ".bson") {
-		baseName = strings.TrimSuffix(baseFileName, ".bson")
+		collName = strings.TrimSuffix(baseFileName, ".bson")
 		fileType = BSONFileType
+		metadataFullPath = strings.TrimSuffix(filename, ".bson") + ".metadata.json"
+	}
+
+	// If the collection name is not truncated, unescape it and return it.
+	if !strings.Contains(collName, "%24") {
+		unescapedCollName, err = util.UnescapeCollectionName(collName)
+		if err != nil {
+			return "", UnknownFileType, fmt.Errorf("error parsing collection name from filename \"%v\": %v", baseFileName, err)
+		}
+		return unescapedCollName, fileType, nil
+	}
+
+	// Otherwise parse it from the metadata file.
+	collName, err = restore.getCollectionNameFromMetadata(metadataFullPath)
+	if err != nil {
+		return "", UnknownFileType, err
+	}
+	unescapedCollName, err = util.UnescapeCollectionName(collName)
+	if err != nil {
+		return "", UnknownFileType, fmt.Errorf("error parsing collection name from filename \"%v\": %v", baseFileName, err)
+	}
+
+	return unescapedCollName, fileType, nil
```

Comment:
Notice that lines 249-253 are the same as lines 26-266. You can refactor this so the conditional checks the positive case and acts on that:
```
if strings.Contains(collName, "%24") {
    collName, err = restore.getCollectionNameFromMetadata(metadataFullPath)
    if err != nil { return "", UnknownFileType, err }
}
... // the duplicated lines I mentioned
```

Reply:
Yes great point. I was wondering why it was looking a lil long the way I had it 🙃

Reply:
I generally prefer conditionals that read as "if true do this" anyway, instead of "if _not_ something, do this"!